# Capstone Project: IBM-Coursera Professional Data Science Specialization



Introduction

This project is designed to answer the question: Is there a San Francisco neighbourhood that has is a good location to open a Chinese restaurant.

This is my project notebook. I have also submitted a report on Github, and have posted this information to my blog.



# 1. Import, install and load.¶

In [4]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library


from bs4 import BeautifulSoup
import urllib.request


print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

In [5]:
!pip install geopy
!conda update folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be UPDATED:

    ca-certificates: 2020.6.20-hecda079_0     conda-forge --> 2020.7.22-0      
    certifi:         2020.6.20-py36h9f0ad1d_0 conda-forge --> 2020.6.20-py36_0 

The following packages will be DOWNGRADED:

    openssl:         1.1.1g-h516909a_1        conda-forge --> 1.1.1g-h7b6447c_0

Preparing transaction: done
Verifying transaction: done
Executing transaction: done


# 2. Find data from the Web page containing San Francisco Zip codes, scrape into Jupyter notebook

In [6]:
#Gets the url and scrapes the html 
url = 'http://ciclt.net/sn/clt/capitolimpact/gw_ziplist.aspx?ClientCode=capitolimpact&State=ca&StName=California&StFIPS=06&FIPS=06075'
req = urllib.request.urlopen(url)


soup = BeautifulSoup(req)

In [7]:
#Finds the table to scrape
table = soup.find('table')

#Provides the empty arrays for the html tags that are being grabbed and assigned to the headings
P = []
C = []

for row in table.find_all('tr'):
    cells = row.find_all('td')
    if len(cells) == 3:
        P.append(cells[0].find(text=True))
        C.append(cells[1].find(text=True))

# 3. Merge Zip code numbers into City data

In [8]:
#Creates the dataframe and places the data in its respective columns
df_sanfran = pd.DataFrame(P, columns=['PostalCode'])
df_sanfran['City'] = C
df_sanfran.head(90)

PostalCode           City
0       94101  San Francisco
1       94102  San Francisco
2       94103  San Francisco
3       94104  San Francisco
4       94105  San Francisco
5       94107  San Francisco
6       94108  San Francisco
7       94109  San Francisco
8       94110  San Francisco
9       94111  San Francisco
10      94112  San Francisco
11      94114  San Francisco
12      94115  San Francisco
13      94116  San Francisco
14      94117  San Francisco
15      94118  San Francisco
16      94119  San Francisco
17      94120  San Francisco
18      94121  San Francisco
19      94122  San Francisco
20      94123  San Francisco
21      94124  San Francisco
22      94125  San Francisco
23      94126  San Francisco
24      94127  San Francisco
25      94129  San Francisco
26      94130  San Francisco
27      94131  San Francisco
28      94132  San Francisco
29      94133  San Francisco
30      94134  San Francisco
31      94140  San Francisco
32      94141  San Francisco
33      94142  San Francisco
34      94146  San Francisco
35      94147  San Francisco
36      94157  San Francisco
37      94159  San Francisco
38      94164  San Francisco
39      94165  San Francisco
40      94166  San Francisco
41      94167  San Francisco
42      94168  San Francisco
43      94169  San Francisco
44      94170  San Francisco
45      94172  San Francisco
46      94188  San Francisco

In [9]:
#A specialized function that joins the neighborhoods with the same postalcode
foo = lambda a: ','.join(a) 
df_sanfran = df_sanfran.groupby(['PostalCode']).agg({
                                'City': foo}).reset_index()

In [10]:
df_sanfran.head(48)

PostalCode           City
0       94101  San Francisco
1       94102  San Francisco
2       94103  San Francisco
3       94104  San Francisco
4       94105  San Francisco
5       94107  San Francisco
6       94108  San Francisco
7       94109  San Francisco
8       94110  San Francisco
9       94111  San Francisco
10      94112  San Francisco
11      94114  San Francisco
12      94115  San Francisco
13      94116  San Francisco
14      94117  San Francisco
15      94118  San Francisco
16      94119  San Francisco
17      94120  San Francisco
18      94121  San Francisco
19      94122  San Francisco
20      94123  San Francisco
21      94124  San Francisco
22      94125  San Francisco
23      94126  San Francisco
24      94127  San Francisco
25      94129  San Francisco
26      94130  San Francisco
27      94131  San Francisco
28      94132  San Francisco
29      94133  San Francisco
30      94134  San Francisco
31      94140  San Francisco
32      94141  San Francisco
33      94142  San Francisco
34      94146  San Francisco
35      94147  San Francisco
36      94157  San Francisco
37      94159  San Francisco
38      94164  San Francisco
39      94165  San Francisco
40      94166  San Francisco
41      94167  San Francisco
42      94168  San Francisco
43      94169  San Francisco
44      94170  San Francisco
45      94172  San Francisco
46      94188  San Francisco

# 4. Read the csv file containing latitude and longitude for each Zip code.

In [11]:
df_lonlat = pd.read_csv('https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/download/?format=csv&refine.state=CA&q=San+Francisco&timezone=America/Los_Angeles&lang=en&use_labels_for_header=true&csv_separator=%3B', delimiter=';')
df_lonlat.head(73)



Zip                 City State   Latitude   Longitude  Timezone  \
0   94138        San Francisco    CA  37.784827 -122.727802        -8   
1   94169        San Francisco    CA  37.784827 -122.727802        -8   
2   94153        San Francisco    CA  37.784827 -122.727802        -8   
3   94146        San Francisco    CA  37.784827 -122.727802        -8   
4   94165        San Francisco    CA  37.784827 -122.727802        -8   
5   94155        San Francisco    CA  37.784827 -122.727802        -8   
6   94080  South San Francisco    CA  37.652857 -122.430100        -8   
7   94175        San Francisco    CA  37.784827 -122.727802        -8   
8   94160        San Francisco    CA  37.784827 -122.727802        -8   
9   94164        San Francisco    CA  37.784827 -122.727802        -8   
10  94131        San Francisco    CA  37.741797 -122.437800        -8   
11  94114        San Francisco    CA  37.758434 -122.435120        -8   
12  94116        San Francisco    CA  37.743381 -122.485780        -8   
13  94177        San Francisco    CA  37.784827 -122.727802        -8   
14  94130        San Francisco    CA  37.819423 -122.369660        -8   
15  94159        San Francisco    CA  37.784827 -122.727802        -8   
16  94167        San Francisco    CA  37.784827 -122.727802        -8   
17  94125        San Francisco    CA  37.784827 -122.727802        -8   
18  94154        San Francisco    CA  37.784827 -122.727802        -8   
19  94137        San Francisco    CA  37.784827 -122.727802        -8   
20  94117        San Francisco    CA  37.770937 -122.442760        -8   
21  94172        San Francisco    CA  37.784827 -122.727802        -8   
22  94104        San Francisco    CA  37.791728 -122.401900        -8   
23  94122        San Francisco    CA  37.758380 -122.484780        -8   
24  94141        San Francisco    CA  37.784827 -122.727802        -8   
25  94110        San Francisco    CA  37.748730 -122.415450        -8   
26  94102        San Francisco    CA  37.779329 -122.419150        -8   
27  94150        San Francisco    CA  37.784827 -122.727802        -8   
28  94112        San Francisco    CA  37.720931 -122.442410        -8   
29  94132        San Francisco    CA  37.724231 -122.479580        -8   
30  94124        San Francisco    CA  37.732797 -122.393480        -8   
31  94134        San Francisco    CA  37.719581 -122.410850        -8   
32  94123        San Francisco    CA  37.801028 -122.438360        -8   
33  94119        San Francisco    CA  37.784827 -122.727802        -8   
34  94135        San Francisco    CA  37.784827 -122.727802        -8   
35  94133        San Francisco    CA  37.801878 -122.410180        -8   
36  94108        San Francisco    CA  37.792678 -122.407930        -8   
37  94107        San Francisco    CA  37.766529 -122.395770        -8   
38  94118        San Francisco    CA  37.782029 -122.461580        -8   
39  94103        San Francisco    CA  37.772329 -122.410870        -8   
40  94162        San Francisco    CA  37.784827 -122.727802        -8   
41  94139        San Francisco    CA  37.784827 -122.727802        -8   
42  94136        San Francisco    CA  37.784827 -122.727802        -8   
43  94168        San Francisco    CA  37.784827 -122.727802        -8   
44  94099  South San Francisco    CA  37.381144 -122.334825        -8   
45  94188        San Francisco    CA  37.784827 -122.727802        -8   
46  94105        San Francisco    CA  37.789228 -122.395700        -8   
47  94161        San Francisco    CA  37.784827 -122.727802        -8   
48  94145        San Francisco    CA  37.784827 -122.727802        -8   
49  94128        San Francisco    CA  37.621964 -122.395340        -8   
50  94166        San Francisco    CA  37.784827 -122.727802        -8   
51  94126        San Francisco    CA  37.784827 -122.727802        -8   
52  94101        San Francisco    CA  37.784827 -122.727802        -8   
53  94143        San Francisco    CA  37.763130 -122.458582     

# 4A. Drop unnecessary columns¶

In [12]:
df_lonlat.drop(['City', 'State', 'Timezone', 'Daylight savings time flag', 'geopoint'], axis=1, inplace=True)
df_lonlat.head()

Zip   Latitude   Longitude
0  94138  37.784827 -122.727802
1  94169  37.784827 -122.727802
2  94153  37.784827 -122.727802
3  94146  37.784827 -122.727802
4  94165  37.784827 -122.727802

In [13]:
df_lonlat.rename(columns={'Zip':'PostalCode'}, inplace=True)


In [14]:
df_sanfran.PostalCode = df_sanfran.PostalCode.astype(int)

# 5. Merge the Zip Code dataframe and the latitude / longitude dataframe.

In [15]:
df_sanfran = pd.merge(df_sanfran, df_lonlat, on='PostalCode', how='outer')


# 6. Generate map showing Zip codes.

In [16]:
# create map of San Francisco using latitude and longitude values
map_sanfran = folium.Map(location=[37.748730, -122.415450], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_sanfran['Latitude'], df_sanfran['Longitude'], df_sanfran['City']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_sanfran)  
    
map_sanfran

# 6A. Scrape data from Web page listing San Francisco Zip codes.

In [17]:
#Gets the url and scrapes the html 
url1 = 'https://www.zipdatamaps.com/zipcodes-san-francisco-ca'
req1 = urllib.request.urlopen(url1)

soup1 = BeautifulSoup(req1)

# 6B. Load data into dataframe.

In [18]:
table1 = soup1.find('table', class_='table')

Post = []
Pop = []

for row in table1.find_all('tr'):
    cells = row.find_all('td')
    if len(cells) == 8:
        Post.append(cells[0].find(text=True))
        Pop.append(cells[5].find(text=True))

# 6C. Merge the Zip code, latitude, and longitude dataframes with population dataframe.

In [19]:
df_pop = pd.DataFrame(Post, columns=['PostalCode'])
df_pop['Population'] = Pop

In [20]:
df_pop.PostalCode = df_pop.PostalCode.astype(int)

In [21]:
df_merged = pd.merge(df_sanfran, df_pop, on='PostalCode', how='outer')


In [22]:
%%debug
cols = [2, 3, 4, 5, ]
df_merged.drop(df_merged.columns[cols], axis=0, inplace=True)


NOTE: Enter 'c' at the ipdb>  prompt to continue execution.
> <string>(2)<module>()

ipdb> c
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/indexes/base.py in __getitem__(self, key)
   3966 
   3967         key = com.values_from_object(key)
-> 3968         result = getitem(key)
   3969         if not is_scalar(result):
   3970             return promote(result)

IndexError: index 5 is out of bounds for axis 1 with size 5


In [23]:


df_merged = df_merged.dropna()

In [24]:
df_merged.head(70)

PostalCode           City   Latitude  Longitude     Population
1        94102  San Francisco  37.779329 -122.41915  San Francisco
2        94103  San Francisco  37.772329 -122.41087  San Francisco
3        94104  San Francisco  37.791728 -122.40190  San Francisco
4        94105  San Francisco  37.789228 -122.39570  San Francisco
5        94107  San Francisco  37.766529 -122.39577  San Francisco
6        94108  San Francisco  37.792678 -122.40793  San Francisco
7        94109  San Francisco  37.792778 -122.42188  San Francisco
8        94110  San Francisco  37.748730 -122.41545  San Francisco
9        94111  San Francisco  37.798228 -122.40027  San Francisco
10       94112  San Francisco  37.720931 -122.44241  San Francisco
11       94114  San Francisco  37.758434 -122.43512  San Francisco
12       94115  San Francisco  37.786129 -122.43736  San Francisco
13       94116  San Francisco  37.743381 -122.48578  San Francisco
14       94117  San Francisco  37.770937 -122.44276  San Francisco
15       94118  San Francisco  37.782029 -122.46158  San Francisco
18       94121  San Francisco  37.778729 -122.49265  San Francisco
19       94122  San Francisco  37.758380 -122.48478  San Francisco
20       94123  San Francisco  37.801028 -122.43836  San Francisco
21       94124  San Francisco  37.732797 -122.39348  San Francisco
24       94127  San Francisco  37.734964 -122.45970  San Francisco
25       94129  San Francisco  37.799840 -122.46167  San Francisco
26       94130  San Francisco  37.819423 -122.36966  San Francisco
27       94131  San Francisco  37.741797 -122.43780  San Francisco
28       94132  San Francisco  37.724231 -122.47958  San Francisco
29       94133  San Francisco  37.801878 -122.41018  San Francisco
30       94134  San Francisco  37.719581 -122.41085  San Francisco

# 7. Submit Foursquare credentials¶

In [25]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: Y5G2ZFFUXD23IOJUOG4UATIMKMDQUJXLLXTAEYAHPTU2SZLL
CLIENT_SECRET:JFUCDQLXDX2E4ZP20EHH5JJ3H4L1QDJRT25ES14J0WA1C1VO


# 8. We look at all San Francisco Zip codes, using 94107 as a reference point

In [26]:
first_nei = df_merged['PostalCode'][5]
first_nei

94107

In [27]:
first_nei_lat = df_merged.loc[5,'Latitude']
first_nei_lon = df_merged.loc[5,'Longitude']
print('Latitude and longitude values of {} are {}, {}.'.format(first_nei, 
                                                               first_nei_lat, 
                                                               first_nei_lon))

Latitude and longitude values of 94107 are 37.766529, -122.39577.


In [28]:
radius = 500 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    first_nei_lat, 
    first_nei_lon, 
    radius, 
    LIMIT)

In [29]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f605bc5bf4f711750d3057d'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'}]},
  'headerLocation': 'Showplace Square',
  'headerFullLocation': 'Showplace Square, San Francisco',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 28,
  'suggestedBounds': {'ne': {'lat': 37.7710290045, 'lng': -122.39008811633391},
   'sw': {'lat': 37.762028995499996, 'lng': -122.40145188366608}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '58d451f495383907dba37158',
       'name': 'Boba Guys',
       'location': {'address': '1002 16th St',
        'crossStreet': 'at Missouri St',
        'lat': 37.76644771569975,
        'lng': -122.39704236067308,
        'labeledLatLngs': 

In [30]:
results['response']['groups'][0]['items'][0]['venue']['categories'][0]['name']

'Bubble Tea Shop'

In [31]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['reasons.count', 'reasons.items', 'referralId', 'venue.categories',
       'venue.delivery.id', 'venue.delivery.provider.icon.name',
       'venue.delivery.provider.icon.prefix',
       'venue.delivery.provider.icon.sizes', 'venue.delivery.provider.name',
       'venue.delivery.url', 'venue.id', 'venue.location.address',
       'venue.location.cc', 'venue.location.city', 'venue.location.country',
       'venue.location.crossStreet', 'venue.location.distance',
       'venue.location.formattedAddress', 'venue.location.labeledLatLngs',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.neighborhood', 'venue.location.postalCode',
       'venue.location.state', 'venue.name', 'venue.photos.count',
       'venue.photos.groups', 'venue.venuePage.id'],
      dtype='object')

# 9. Next, we gather information about the venues in Zip Code 94114 from Foursquare.

In [32]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [33]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]
nearby_venues

venue.name  \
0                                  Boba Guys   
1                         Bottom of the Hill   
2                             Fitness Urbano   
3   Pawtrero Hill Bathhouse and Feed Company   
4            UCSF Bakar Fitness & Rec Center   
5                              Daggett Plaza   
6                       Big Daddy's Antiques   
7        Trimark Economy Restaurant Fixtures   
8              Intelligentsia Roasting Works   
9                               Seven Stills   
10                                 Ruby Wine   
11                    UCSF Bakar Center Pool   
12                                      Plow   
13                        Connecticut Yankee   
14                           Chez Maman East   
15                             Thee Parkside   
16                           Hazel's Kitchen   
17                                  Farley's   
18                               Parker Cafe   
19                 Jackson Park & Playground   
20                          Provender Coffee   
21                       Christopher's Books   
22                                    Papito   
23                                   Mochica   
24                                 Sunflower   
25                            The Wine House   
26                        Cracked & Battered   
27                           Goat Hill Pizza   

                                     venue.categories  venue.location.lat  \
0   [{'id': '52e81612bcbc57f1066b7a0c', 'name': 'B...           37.766448   
1   [{'id': '4bf58dd8d48988d1e9931735', 'name': 'R...           37.765116   
2   [{'id': '4bf58dd8d48988d175941735', 'name': 'G...           37.765684   
3   [{'id': '4bf58dd8d48988d100951735', 'name': 'P...           37.764140   
4   [{'id': '4bf58dd8d48988d176941735', 'name': 'G...           37.768146   
5   [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           37.766920   
6   [{'id': '4bf58dd8d48988d116951735', 'name': 'A...           37.765015   
7   [{'id': '4bf58dd8d48988d1ff941735', 'name': 'M...           37.767543   
8   [{'id': '4bf58dd8d48988d124941735', 'name': 'O...           37.763744   
9   [{'id': '4e0e22f5a56208c4ea9a85a0', 'name': 'D...           37.768620   
10  [{'id': '4bf58dd8d48988d119951735', 'name': 'W...           37.762541   
11  [{'id': '4bf58dd8d48988d15e941735', 'name': 'P...           37.768232   
12  [{'id': '4bf58dd8d48988d143941735', 'name': 'B...           37.762516   
13  [{'id': '4bf58dd8d48988d14e941735', 'name': 'A...           37.765005   
14  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...           37.762531   
15  [{'id': '4bf58dd8d48988d118941735', 'name': 'D...           37.765067   
16  [{'id': '4bf58dd8d48988d1c5941735', 'name': 'S...           37.762633   
17  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           37.762596   
18  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...           37.762555   
19  [{'id': '4bf58dd8d48988d163941735', 'name': 'P...           37.764303   
20  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...           37.762456   
21  [{'id': '4bf58dd8d48988d114951735', 'name': 'B...           37.762585   
22  [{'id': '4bf58dd8d48988d1c1941735', 'name': 'M...           37.762346   
23  [{'id': '4eb1bfa43b7b52c0e1adc2e8', 'name': 'P...           37.762467   
24  [{'id': '4bf58dd8d48988d14a941735', 'name': 'V...           37.762609   
25  [{'id': '4bf58dd8d48988d119951735', 'name': 'W...           37.765602   
26  [{'id': '4bf58dd8d48988d143941735', 'name': 'B...           37.762583   
27  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...           37.762434   

    venue.location.lng  
0          -122.397042  
1          -122.396218  
2          -122.397009  
3          -122.394500  
4          -122.393290  
5          -122.396027  
6          -122.399497  
7          -122.397641  
8          -122.395227  
9          -122.399168  
10         -122.397027  
11         -122.393217  
12         -122.395467  
13         -122.397903  
14         -122.396652  
15     

In [34]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name             categories  \
0                                  Boba Guys        Bubble Tea Shop   
1                         Bottom of the Hill              Rock Club   
2                             Fitness Urbano   Gym / Fitness Center   
3   Pawtrero Hill Bathhouse and Feed Company              Pet Store   
4            UCSF Bakar Fitness & Rec Center                    Gym   
5                              Daggett Plaza                   Park   
6                       Big Daddy's Antiques           Antique Shop   
7        Trimark Economy Restaurant Fixtures     Miscellaneous Shop   
8              Intelligentsia Roasting Works                 Office   
9                               Seven Stills             Distillery   
10                                 Ruby Wine              Wine Shop   
11                    UCSF Bakar Center Pool                   Pool   
12                                      Plow         Breakfast Spot   
13                        Connecticut Yankee    American Restaurant   
14                           Chez Maman East      French Restaurant   
15                             Thee Parkside               Dive Bar   
16                           Hazel's Kitchen         Sandwich Place   
17                                  Farley's            Coffee Shop   
18                               Parker Cafe                   Café   
19                 Jackson Park & Playground                   Park   
20                          Provender Coffee                   Café   
21                       Christopher's Books              Bookstore   
22                                    Papito     Mexican Restaurant   
23                                   Mochica    Peruvian Restaurant   
24                                 Sunflower  Vietnamese Restaurant   
25                            The Wine House              Wine Shop   
26                        Cracked & Battered         Breakfast Spot   
27                           Goat Hill Pizza            Pizza Place   

          lat         lng  
0   37.766448 -122.397042  
1   37.765116 -122.396218  
2   37.765684 -122.397009  
3   37.764140 -122.394500  
4   37.768146 -122.393290  
5   37.766920 -122.396027  
6   37.765015 -122.399497  
7   37.767543 -122.397641  
8   37.763744 -122.395227  
9   37.768620 -122.399168  
10  37.762541 -122.397027  
11  37.768232 -122.393217  
12  37.762516 -122.395467  
13  37.765005 -122.397903  
14  37.762531 -122.396652  
15  37.765067 -122.399842  
16  37.762633 -122.396058  
17  37.762596 -122.395984  
18  37.762555 -122.396420  
19  37.764303 -122.399572  
20  37.762456 -122.396839  
21  37.762585 -122.396605  
22  37.762346 -122.397409  
23  37.762467 -122.397298  
24  37.762609 -122.397538  
25  37.765602 -122.400673  
26  37.762583 -122.397378  
27  37.762434 -122.397654


# 10. We retrieve lists of venues in each San Francisco Zip code.

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
sanfran_venues = getNearbyVenues(names = df_merged['PostalCode'],
                                   latitudes = df_merged['Latitude'],
                                   longitudes = df_merged['Longitude']
                                  )



94102
94103
94104
94105
94107
94108
94109
94110
94111
94112
94114
94115
94116
94117
94118
94121
94122
94123
94124
94127
94129
94130
94131
94132
94133
94134


In [37]:
print(sanfran_venues.shape)
sanfran_venues.head(70)

(1417, 7)


PostalCode  Neighborhood Latitude  Neighborhood Longitude  \
0        94102              37.779329              -122.41915   
1        94102              37.779329              -122.41915   
2        94102              37.779329              -122.41915   
3        94102              37.779329              -122.41915   
4        94102              37.779329              -122.41915   
5        94102              37.779329              -122.41915   
6        94102              37.779329              -122.41915   
7        94102              37.779329              -122.41915   
8        94102              37.779329              -122.41915   
9        94102              37.779329              -122.41915   
10       94102              37.779329              -122.41915   
11       94102              37.779329              -122.41915   
12       94102              37.779329              -122.41915   
13       94102              37.779329              -122.41915   
14       94102              37.779329              -122.41915   
15       94102              37.779329              -122.41915   
16       94102              37.779329              -122.41915   
17       94102              37.779329              -122.41915   
18       94102              37.779329              -122.41915   
19       94102              37.779329              -122.41915   
20       94102              37.779329              -122.41915   
21       94102              37.779329              -122.41915   
22       94102              37.779329              -122.41915   
23       94102              37.779329              -122.41915   
24       94102              37.779329              -122.41915   
25       94102              37.779329              -122.41915   
26       94102              37.779329              -122.41915   
27       94102              37.779329              -122.41915   
28       94102              37.779329              -122.41915   
29       94102              37.779329              -122.41915   
30       94102              37.779329              -122.41915   
31       94102              37.779329              -122.41915   
32       94102              37.779329              -122.41915   
33       94102              37.779329              -122.41915   
34       94102              37.779329              -122.41915   
35       94102              37.779329              -122.41915   
36       94102              37.779329              -122.41915   
37       94102              37.779329              -122.41915   
38       94102              37.779329              -122.41915   
39       94102              37.779329              -122.41915   
40       94102              37.779329              -122.41915   
41       94102              37.779329              -122.41915   
42       94102              37.779329              -122.41915   
43       94102              37.779329              -122.41915   
44       94102              37.779329              -122.41915   
45       94102              37.779329              -122.41915   
46       94102              37.779329              -122.41915   
47       94102              37.779329              -122.41915   
48       94102              37.779329              -122.41915   
49       94102              37.779329              -122.41915   
50       94102              37.779329              -122.41915   
51       94102              37.779329              -122.41915   
52       94102              37.779329              -122.41915   
53       94102              37.779329              -122.41915   
54       94102              37.779329              -122.41915   
55       94102              37.779329              -122.41915   
56       94102              37.779329              -122.41915   
57       94102              37.779329              -122.41915   
58       94102              37.779329              -122.41915   
59       94102              37.779329              -122.41915   
60       94102              37.779329  

In [38]:
sanfran_venues.groupby('PostalCode').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
PostalCode                                                         
94102                          90                      90     90   
94103                          77                      77     77   
94104                         100                     100    100   
94105                          81                      81     81   
94107                          28                      28     28   
94108                          93                      93     93   
94109                          85                      85     85   
94110                          47                      47     47   
94111                         100                     100    100   
94112                          30                      30     30   
94114                          79                      79     79   
94115                          99                      99     99   
94116                          43                      43     43   
94117                          38                      38     38   
94118                          55                      55     55   
94121                          44                      44     44   
94122                          10                      10     10   
94123                          84                      84     84   
94124                          22                      22     22   
94127                           2                       2      2   
94129                          29                      29     29   
94130                          11                      11     11   
94131                          12                      12     12   
94132                          54                      54     54   
94133                         100                     100    100   
94134                           4                       4      4   

            Venue Latitude  Venue Longitude  Venue Category  
PostalCode                                                   
94102                   90               90              90  
94103                   77               77              77  
94104                  100              100             100  
94105                   81               81              81  
94107                   28               28              28  
94108                   93               93              93  
94109                   85               85              85  
94110                   47               47              47  
94111                  100              100             100  
94112                   30               30              30  
94114                   79               79              79  
94115                   99               99              99  
94116                   43               43              43  
94117                   38               38              38  
94118                   55               55              55  
94121                   44               44              44  
94122                   10               10              10  
94123                   84               84              84  
94124                   22               22              22  
94127                    2                2               2  
94129                   29               29              29  
94130                   11               11              11  
94131                   12               12              12  
94132                   54               54              54  
94133                  100              100             100  
94134                    4                4               4

In [39]:
print('There are {} unique categories.'.format(len(sanfran_venues['Venue Category'].unique())))

There are 259 unique categories.


In [40]:
sanfran_onehot = pd.get_dummies(sanfran_venues[['Venue Category']], prefix="", prefix_sep="")

sanfran_onehot['PostalCode'] = sanfran_venues['PostalCode'] 
fixed_columns = [sanfran_onehot.columns[-1]] + list(sanfran_onehot.columns[:-1])
sanfran_onehot = sanfran_onehot[fixed_columns]
sanfran_onehot.head(70)


PostalCode  ATM  Acai House  Accessories Store  Adult Boutique  \
0        94102    0           0                  0               0   
1        94102    0           0                  0               0   
2        94102    0           0                  0               0   
3        94102    0           0                  0               0   
4        94102    0           0                  0               0   
5        94102    0           0                  0               0   
6        94102    0           0                  0               0   
7        94102    0           0                  0               0   
8        94102    0           0                  0               0   
9        94102    0           0                  0               0   
10       94102    0           0                  0               0   
11       94102    0           0                  0               0   
12       94102    0           0                  0               0   
13       94102    0           0                  0               0   
14       94102    0           0                  0               0   
15       94102    0           0                  0               0   
16       94102    0           0                  0               0   
17       94102    0           0                  0               0   
18       94102    0           0                  0               0   
19       94102    0           0                  0               0   
20       94102    0           0                  0               0   
21       94102    0           0                  0               0   
22       94102    0           0                  0               0   
23       94102    0           0                  0               0   
24       94102    0           0                  0               0   
25       94102    0           0                  0               0   
26       94102    0           0                  0               0   
27       94102    0           0                  0               0   
28       94102    0           0                  0               0   
29       94102    0           0                  0               0   
30       94102    0           0                  0               0   
31       94102    0           0                  0               0   
32       94102    0           0                  0               0   
33       94102    0           0                  0               0   
34       94102    0           0                  0               0   
35       94102    0           0                  0               0   
36       94102    0           0                  0               0   
37       94102    0           0                  0               0   
38       94102    0           0                  0               0   
39       94102    0           0                  0               0   
40       94102    0           0                  0               0   
41       94102    0           0                  0               0   
42       94102    0           0                  0               0   
43       94102    0           0                  0               0   
44       94102    0           0                  0               0   
45       94102    0           0                  0               0   
46       94102    0           0                  0               0   
47       94102    0           0                  0               0   
48       94102    0           0                  0               0   
49       94102    0           0                  0               0   
50       94102    0           0                  0               0   
51       94102    0           0                  0               0   
52       94102    0           0                  0               0   
53       94102    0           0                  0               0   
54       94102    0           0                  0               0   
55       94102    0           0                  0               0   
56       94102

In [41]:
sanfran_grouped = sanfran_onehot.groupby('PostalCode').mean().reset_index()
sanfran_grouped

PostalCode       ATM  Acai House  Accessories Store  Adult Boutique  \
0        94102  0.000000        0.00           0.000000        0.000000   
1        94103  0.000000        0.00           0.000000        0.000000   
2        94104  0.000000        0.01           0.000000        0.000000   
3        94105  0.000000        0.00           0.000000        0.000000   
4        94107  0.000000        0.00           0.000000        0.000000   
5        94108  0.000000        0.00           0.000000        0.000000   
6        94109  0.000000        0.00           0.000000        0.000000   
7        94110  0.000000        0.00           0.000000        0.000000   
8        94111  0.000000        0.01           0.000000        0.000000   
9        94112  0.000000        0.00           0.000000        0.000000   
10       94114  0.000000        0.00           0.000000        0.012658   
11       94115  0.000000        0.00           0.000000        0.000000   
12       94116  0.000000        0.00           0.000000        0.000000   
13       94117  0.000000        0.00           0.026316        0.000000   
14       94118  0.018182        0.00           0.018182        0.000000   
15       94121  0.000000        0.00           0.000000        0.000000   
16       94122  0.000000        0.00           0.000000        0.000000   
17       94123  0.000000        0.00           0.000000        0.000000   
18       94124  0.000000        0.00           0.000000        0.000000   
19       94127  0.000000        0.00           0.000000        0.000000   
20       94129  0.000000        0.00           0.000000        0.000000   
21       94130  0.000000        0.00           0.000000        0.000000   
22       94131  0.000000        0.00           0.000000        0.000000   
23       94132  0.000000        0.00           0.018519        0.000000   
24       94133  0.000000        0.00           0.010000        0.000000   
25       94134  0.000000        0.00           0.000000        0.000000   

    African Restaurant  Alternative Healer  American Restaurant  Antique Shop  \
0             0.000000            0.000000             0.011111      0.000000   
1             0.000000            0.000000             0.000000      0.000000   
2             0.000000            0.000000             0.010000      0.000000   
3             0.000000            0.012346             0.000000      0.000000   
4             0.000000            0.000000             0.035714      0.035714   
5             0.000000            0.000000             0.010753      0.000000   
6             0.023529            0.000000             0.023529      0.000000   
7             0.000000            0.000000             0.000000      0.000000   
8             0.000000            0.000000             0.020000      0.000000   
9             0.000000            0.000000             0.000000      0.000000   
10            0.000000            0.000000             0.012658      0.000000   
11            0.000000            0.000000             0.030303      0.000000   
12            0.000000            0.000000             0.000000      0.000000   
13            0.000000            0.000000             0.000000      0.000000   
14            0.000000            0.000000             0.018182      0.000000   
15            0.000000            0.000000             0.045455      0.022727   
16            0.000000            0.000000             0.000000      0.000000   
17            0.000000            0.011905             0.023810      0.000000   
18            0.045455            0.000000             0.000000      0.000000   
19            0.000000            0.000000             0.000000      0.000000   
20            0.000000            0.000000             0.034483      0.000000   
21            0.000000            0.000000             0.000000      0.000000   
22            0.000000            0.000000             0.000000      0.000000   
23            0.000000            

In [42]:
sanfran_grouped.PostalCode = sanfran_grouped.PostalCode.astype(str)

# 11. Identify the most common venues for each Zip code.

In [43]:
num_top_venues = 5
for code in sanfran_grouped['PostalCode']:
    print("----"+code+"----")
    temp = sanfran_grouped[sanfran_grouped['PostalCode'] == code].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----94102----
                           venue  freq
0                    Coffee Shop  0.06
1                          Hotel  0.04
2                           Café  0.04
3              French Restaurant  0.03
4  Vegetarian / Vegan Restaurant  0.03


----94103----
             venue  freq
0        Nightclub  0.09
1          Gay Bar  0.05
2     Cocktail Bar  0.05
3              Bar  0.04
4  Thai Restaurant  0.04


----94104----
                 venue  freq
0          Coffee Shop  0.07
1           Food Truck  0.05
2          Men's Store  0.04
3  Japanese Restaurant  0.04
4       Sandwich Place  0.03


----94105----
                  venue  freq
0           Coffee Shop  0.16
1            Food Truck  0.05
2                  Café  0.05
3                   Gym  0.04
4  Gym / Fitness Center  0.04


----94107----
            venue  freq
0            Café  0.07
1       Wine Shop  0.07
2            Park  0.07
3  Breakfast Spot  0.07
4            Pool  0.04


----94108----
                venue  f

In [44]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
sanfran_grouped.PostalCode = sanfran_grouped.PostalCode.astype(int)

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = sanfran_grouped['PostalCode']

for ind in np.arange(sanfran_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(sanfran_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

PostalCode            1st Most Common Venue 2nd Most Common Venue  \
0        94102                      Coffee Shop                 Hotel   
1        94103                        Nightclub          Cocktail Bar   
2        94104                      Coffee Shop            Food Truck   
3        94105                      Coffee Shop            Food Truck   
4        94107                        Wine Shop        Breakfast Spot   
5        94108                            Hotel                Bakery   
6        94109                    Grocery Store  Gym / Fitness Center   
7        94110               Mexican Restaurant           Pizza Place   
8        94111                             Café            Food Truck   
9        94112                      Pizza Place           Bus Station   
10       94114                          Gay Bar       Thai Restaurant   
11       94115                           Bakery                   Spa   
12       94116               Chinese Restaurant    Light Rail Station   
13       94117                      Coffee Shop              Boutique   
14       94118               Chinese Restaurant                Bakery   
15       94121                             Café    Chinese Restaurant   
16       94122               Chinese Restaurant            Playground   
17       94123               Italian Restaurant  Gym / Fitness Center   
18       94124  Southern / Soul Food Restaurant    Mexican Restaurant   
19       94127                            Trail              Bus Line   
20       94129                       Food Truck                Museum   
21       94130                    Grocery Store           Music Venue   
22       94131                             Park                 Trail   
23       94132                   Sandwich Place                  Café   
24       94133                      Coffee Shop           Pizza Place   
25       94134                   Baseball Field                Garden   

     3rd Most Common Venue  4th Most Common Venue  \
0                     Café               Wine Bar   
1                  Gay Bar             Food Truck   
2              Men's Store    Japanese Restaurant   
3                     Café         Sandwich Place   
4                     Park                   Café   
5              Coffee Shop     Chinese Restaurant   
6            Deli / Bodega       Sushi Restaurant   
7              Coffee Shop                   Park   
8           Scenic Lookout            Men's Store   
9       Mexican Restaurant  Vietnamese Restaurant   
10             Coffee Shop            Yoga Studio   
11                    Café         Cosmetics Shop   
12     Dumpling Restaurant                    Bar   
13  Thrift / Vintage Store                   Park   
14      Burmese Restaurant            Pizza Place   
15       Convenience Store            Pizza Place   
16           Train Station           Dessert Shop   
17          Sandwich Place      French Restaurant   
18                    Café                 Bakery   
19             Yoga Studio            Flea Market   
20                    Café            Yoga Studio   
21             Flea Market        Harbor / Marina   
22           Grocery Store          Shopping Mall   
23          Cosmetics Shop              Juice Bar   
24      Italian Restaurant                   Café   
25                   Trail                   Park   

            5th Most Common Venue  6th Most Common Venue  \
0   Vegetarian / Vegan Restaurant      French Restaurant   
1                 Motorcycle Shop                    Bar   
2                  Sandwich Place       Sushi Restaurant   
3                     Art Gallery   Gym / Fitness Center   
4             Peruvian Restaurant                   Pool   
5                     Men's Store           Cocktail Bar   
6                        Wine Bar  Vietnamese Restaurant   
7                        Dive Bar   Gym / Fitness Center   
8              Italian Restaurant         Cosmetics Shop   


# 12. Identify clusters within the data.

In [46]:
from sklearn.cluster import KMeans

kclusters = 5

sanfran_grouped_clustering = sanfran_grouped.drop('PostalCode', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(sanfran_grouped_clustering)

kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 4,
       3, 0, 0, 2], dtype=int32)

In [47]:
sanfran_merged = df_merged[0:47]

sanfran_merged['Cluster Labels'] = kmeans.labels_

sanfran_merged = sanfran_merged.join(neighborhoods_venues_sorted.set_index('PostalCode'), on='PostalCode')

sanfran_merged.head(47)

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


PostalCode           City   Latitude  Longitude     Population  \
1        94102  San Francisco  37.779329 -122.41915  San Francisco   
2        94103  San Francisco  37.772329 -122.41087  San Francisco   
3        94104  San Francisco  37.791728 -122.40190  San Francisco   
4        94105  San Francisco  37.789228 -122.39570  San Francisco   
5        94107  San Francisco  37.766529 -122.39577  San Francisco   
6        94108  San Francisco  37.792678 -122.40793  San Francisco   
7        94109  San Francisco  37.792778 -122.42188  San Francisco   
8        94110  San Francisco  37.748730 -122.41545  San Francisco   
9        94111  San Francisco  37.798228 -122.40027  San Francisco   
10       94112  San Francisco  37.720931 -122.44241  San Francisco   
11       94114  San Francisco  37.758434 -122.43512  San Francisco   
12       94115  San Francisco  37.786129 -122.43736  San Francisco   
13       94116  San Francisco  37.743381 -122.48578  San Francisco   
14       94117  San Francisco  37.770937 -122.44276  San Francisco   
15       94118  San Francisco  37.782029 -122.46158  San Francisco   
18       94121  San Francisco  37.778729 -122.49265  San Francisco   
19       94122  San Francisco  37.758380 -122.48478  San Francisco   
20       94123  San Francisco  37.801028 -122.43836  San Francisco   
21       94124  San Francisco  37.732797 -122.39348  San Francisco   
24       94127  San Francisco  37.734964 -122.45970  San Francisco   
25       94129  San Francisco  37.799840 -122.46167  San Francisco   
26       94130  San Francisco  37.819423 -122.36966  San Francisco   
27       94131  San Francisco  37.741797 -122.43780  San Francisco   
28       94132  San Francisco  37.724231 -122.47958  San Francisco   
29       94133  San Francisco  37.801878 -122.41018  San Francisco   
30       94134  San Francisco  37.719581 -122.41085  San Francisco   

    Cluster Labels            1st Most Common Venue 2nd Most Common Venue  \
1                0                      Coffee Shop                 Hotel   
2                0                        Nightclub          Cocktail Bar   
3                0                      Coffee Shop            Food Truck   
4                0                      Coffee Shop            Food Truck   
5                0                        Wine Shop        Breakfast Spot   
6                0                            Hotel                Bakery   
7                0                    Grocery Store  Gym / Fitness Center   
8                0               Mexican Restaurant           Pizza Place   
9                0                             Café            Food Truck   
10               0                      Pizza Place           Bus Station   
11               0                          Gay Bar       Thai Restaurant   
12               0                           Bakery                   Spa   
13               0               Chinese Restaurant    Light Rail Station   
14               0                      Coffee Shop              Boutique   
15               0               Chinese Restaurant                Bakery   
18               0                             Café    Chinese Restaurant   
19               0               Chinese Restaurant            Playground   
20               0               Italian Restaurant  Gym / Fitness Center   
21               0  Southern / Soul Food Restaurant    Mexican Restaurant   
24               1                            Trail              Bus Line   
25               0                       Food Truck                Museum   
26               4                    Grocery Store           Music Venue   
27               3                             Park                 Trail   
28               0                   Sandwich Place                  Café   
29               0                      Coffee Shop           Pizza Place   
30               2                   Baseball Field                Garden   

     3rd Most Common Venue  4th M

# 13. Generate map with clusters superimposed.

In [48]:
map_clusters = folium.Map(location=[37.801878, -122.41018], zoom_start=11)

x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)

markers_colors = []
for lat, lon, nei , cluster in zip(sanfran_merged['Latitude'], sanfran_merged['Longitude'], sanfran_merged['PostalCode'], sanfran_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']


In [49]:
sanfran_merged.loc[sanfran_merged['Cluster Labels'] == 0, sanfran_merged.columns[[0] + list(range(5, sanfran_merged.shape[1]))]]

PostalCode  Cluster Labels            1st Most Common Venue  \
1        94102               0                      Coffee Shop   
2        94103               0                        Nightclub   
3        94104               0                      Coffee Shop   
4        94105               0                      Coffee Shop   
5        94107               0                        Wine Shop   
6        94108               0                            Hotel   
7        94109               0                    Grocery Store   
8        94110               0               Mexican Restaurant   
9        94111               0                             Café   
10       94112               0                      Pizza Place   
11       94114               0                          Gay Bar   
12       94115               0                           Bakery   
13       94116               0               Chinese Restaurant   
14       94117               0                      Coffee Shop   
15       94118               0               Chinese Restaurant   
18       94121               0                             Café   
19       94122               0               Chinese Restaurant   
20       94123               0               Italian Restaurant   
21       94124               0  Southern / Soul Food Restaurant   
25       94129               0                       Food Truck   
28       94132               0                   Sandwich Place   
29       94133               0                      Coffee Shop   

   2nd Most Common Venue   3rd Most Common Venue  4th Most Common Venue  \
1                  Hotel                    Café               Wine Bar   
2           Cocktail Bar                 Gay Bar             Food Truck   
3             Food Truck             Men's Store    Japanese Restaurant   
4             Food Truck                    Café         Sandwich Place   
5         Breakfast Spot                    Park                   Café   
6                 Bakery             Coffee Shop     Chinese Restaurant   
7   Gym / Fitness Center           Deli / Bodega       Sushi Restaurant   
8            Pizza Place             Coffee Shop                   Park   
9             Food Truck          Scenic Lookout            Men's Store   
10           Bus Station      Mexican Restaurant  Vietnamese Restaurant   
11       Thai Restaurant             Coffee Shop            Yoga Studio   
12                   Spa                    Café         Cosmetics Shop   
13    Light Rail Station     Dumpling Restaurant                    Bar   
14              Boutique  Thrift / Vintage Store                   Park   
15                Bakery      Burmese Restaurant            Pizza Place   
18    Chinese Restaurant       Convenience Store            Pizza Place   
19            Playground           Train Station           Dessert Shop   
20  Gym / Fitness Center          Sandwich Place      French Restaurant   
21    Mexican Restaurant                    Café                 Bakery   
25                Museum                    Café            Yoga Studio   
28                  Café          Cosmetics Shop              Juice Bar   
29           Pizza Place      Italian Restaurant                   Café   

            5th Most Common Venue  6th Most Common Venue  \
1   Vegetarian / Vegan Restaurant      French Restaurant   
2                 Motorcycle Shop                    Bar   
3                  Sandwich Place       Sushi Restaurant   
4                     Art Gallery   Gym / Fitness Center   
5             Peruvian Restaurant                   Pool   
6                     Men's Store           Cocktail Bar   
7                        Wine Bar  Vietnamese Restaurant   
8                        Dive Bar   Gym / Fitness Center   
9              Italian Restaurant         Cosmetics Shop   
10                 Sandwich Place            Flower Shop   
11                           Café      Indian Restaurant   
12            Am

In [50]:
sanfran_merged.loc[sanfran_merged['Cluster Labels'] == 1, sanfran_merged.columns[[0] + list(range(5, sanfran_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
24       94127               1                 Trail              Bus Line   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
24           Yoga Studio           Flea Market        Farmers Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
24  Fast Food Restaurant                 Field   Filipino Restaurant   

   9th Most Common Venue 10th Most Common Venue  
24           Fish Market            Flower Shop

In [51]:
sanfran_merged.loc[sanfran_merged['Cluster Labels'] == 2, sanfran_merged.columns[[0] + list(range(5, sanfran_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
30       94134               2        Baseball Field                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
30                 Trail                  Park           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
30   Filipino Restaurant        Farmers Market  Fast Food Restaurant   

   9th Most Common Venue 10th Most Common Venue  
30                 Field            Fish Market

In [52]:
sanfran_merged.loc[sanfran_merged['Cluster Labels'] == 3, sanfran_merged.columns[[0] + list(range(5, sanfran_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
27       94131               3                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
27         Grocery Store         Shopping Mall    Salon / Barbershop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
27           Coffee Shop            Playground              Pharmacy   

   9th Most Common Venue 10th Most Common Venue  
27    Dim Sum Restaurant               Dive Bar

In [53]:
sanfran_merged.loc[sanfran_merged['Cluster Labels'] == 4, sanfran_merged.columns[[0] + list(range(5, sanfran_merged.shape[1]))]]

PostalCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
26       94130               4         Grocery Store           Music Venue   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
26           Flea Market       Harbor / Marina    Athletics & Sports   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
26        History Museum   Fried Chicken Joint               Brewery   

   9th Most Common Venue 10th Most Common Venue  
26           Rugby Pitch            Flower Shop